litellm/caching.py

```diff
@@ -1446,12 +1449,389 @@ def delete_cache(self, key):
+                )
+
+        else:
+            print_verbose(
+                "dual-semantic-cache (sync) get_cache: in_memory_cache is None"
+            )
+
+        from redisvl.query import VectorQuery
+
+        messages = kwargs["messages"]
+        prompt = "".join(message["content"] for message in messages)
+
+        embedding_response = litellm.embedding(
+            model=self.embedding_model_name,
+            input=prompt,
+            cache={"no-store": True, "no-cache": True},
+        )
+
+        # get the embedding
+        embedding = embedding_response["data"][0]["embedding"]
+
+        query = VectorQuery(
+            vector=embedding,
+            vector_field_name="litellm_embedding",
+            return_fields=["response", "prompt", "vector_distance"],
+            num_results=1,
+        )
+
+        results = self.index.query(query)
+        if results is None:
+            return None
+        if isinstance(results, list):
+            if len(results) == 0:
+                return None
+
+        vector_distance = results[0]["vector_distance"]
+        vector_distance = float(vector_distance)
+        similarity = 1 - vector_distance
+        cached_prompt = results[0]["prompt"]
+
+        # check similarity, if more than self.similarity_threshold, return results
+        print_verbose(
+            f"dual-semantic-cache (sync): similarity threshold: {self.similarity_threshold}, similarity: {similarity}, prompt: {prompt}, closest_cached_prompt: {cached_prompt}"
+        )
+        if similarity > self.similarity_threshold:
+            cached_value = results[0]["response"]
+            print_verbose(
+                f"dual-semantic-cache (sync) cache hit, similarity: {similarity}, Current prompt: {prompt}, cached_prompt: {cached_prompt}"
+            )
+            return self._get_cache_logic(cached_response=cached_value)
+        else:
+            return None
+
+    async def async_set_cache(self, key, value: dict, **kwargs):
+        import numpy as np
+        from litellm.proxy.proxy_server import llm_router, llm_model_list
+
+        try:
+            await self.index.acreate(overwrite=False)
+
+        except Exception as e:
+            print_verbose(
+                f"dual-semantic-cache (async): exception creating semantic cache index: {str(e)}"
+            )
+            return
+
+        print_verbose(f"dual-semantic-cache (async) set_cache, kwargs: {kwargs}")
+
+        messages = kwargs["messages"]
+        prompt = "".join(message["content"] for message in messages)
+
+        router_model_names = (
+            [m["model_name"] for m in llm_model_list]
+            if llm_model_list is not None
+            else []
+        )
+        if llm_router is not None and self.embedding_model_name in router_model_names:
+            user_api_key = kwargs.get("metadata", {}).get("user_api_key", "")
+            embedding_response = await llm_router.aembedding(
```

Comment:
why switch to using the router here?

Reply:
to be honest i'm not sure what the best practice regarding router usage is - i used it because the redis semantic cache implementation does, but i'm also happy to not use it. what would you recommend?